docs/src/proposals/program-instruction-macro.md

```diff
@@ -0,0 +1,224 @@
+# Program Instruction Macro
+
+## Problem
+
+Currently, inspecting an on-chain transaction requires depending on a
+client-side, language-specific decoding library to parse the instruction.  If
+rpc methods could return decoded instruction details, these custom solutions
+would be unnecessary.
+
+We can deserialize instruction data using a program's Instruction enum, but
+decoding the account-key list into human-readable identifiers requires manual
+parsing. Our current Instruction enums have that account information, but only
+in variant docs.
+
+Similarly, we have instruction constructor functions that duplicate nearly all
+the information in the enum, but we can't generate that constructor from the
+enum definition because the list of account references is in code comments.
+
+Also, Instruction docs can vary between implementations, as there is no
+mechanism to ensure consistency.
+
+## Proposed Solution
+
+Move the data from code comments to attributes, such that the constructors
+can be generated, and include all the documentation from the enum definition.
+
+Here is an example of an Instruction enum using the new accounts format:
+
+```rust,ignore
+#[instructions(test_program::id())]
+pub enum TestInstruction {
+    /// Transfer lamports
+    #[accounts(
+        from_account(signer, writable, desc = "Funding account"),
+        to_account(writable, desc = "Recipient account"),
+    )]
+    Transfer {
+        lamports: u64,
+    },
+
+    /// Provide M of N required signatures
+    #[accounts(
+        data_account(writable, desc = "Data account"),
+        signer(signer, multiple, desc = "Signer"),
+    )]
+    Multisig,
+
+    /// Consumes a stored nonce, replacing it with a successor
+    #[accounts(
+        nonce_account(signer, writable, desc = "Nonce account"),
+        recent_blockhashes_sysvar(signer, writable, desc = "RecentBlockhashes sysvar"),
+        nonce_authority(signer, optional, desc = "Nonce authority"),
+    )]
+    AdvanceNonceAccount,
+}
+```
+
+An example of the generated TestInstruction with docs:
+```rust,ignore
+    pub enum TestInstruction {
+      /// Transfer lamports
+      ///
+      /// * Accounts expected by this instruction:
+      ///   0. `[writable, signer]` Funding account
+      ///   1. `[writable]` Recipient account
+      Transfer {
+          lamports: u64,
+      },
+
+      /// Provide M of N required signatures
+      ///
+      /// * Accounts expected by this instruction:
+      ///   0. `[writable]` Data account
+      ///   * (Multiple) `[signer]` Signers
```

Comment:
Does that `*` get rendered well by `cargo doc`?

Reply:
It stacks an unordered list underneath the ordered list. I think it looks not-terrible:

![image](https://user-images.githubusercontent.com/3030561/85748843-9605da80-b6c5-11ea-91a4-e5fd64955d8f.png)

Reply:
Better than a big header line in the middle of the description :-)